### PR TITLE
http: fix segfault discovered during Lyft's smoketest of #1267.

### PR DESCRIPTION
**Pull request by htuch**
Turns out that a sendLocalReply() will unconditionally send a body after errors such as 504, gRPC
client performs a reset before this and had self-destructed already.

### CODE REVIEW

**Comment by htuch** [comment]
@mattklein123 This should fix your segfault. I'd appreciate thoughts on whether we should also guard in `encode{Headers,Trailers}()`, as the fix seems fragile (it depends entirely on what router is doing in its internal implementation). I originally had stream closed guards for these, but couldn't see an easy way to get code coverage via tests.

File: source/common/http/async_client_impl.cc [hunk 1]
```diff
@@ -79,6 +79,9 @@ void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
       },
       nullptr);
 #endif
+  if (remote_closed_) {
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: How can this one happen? I can see how data can happen. I think trailers can also happen. We need to guard against that also I think.

File: source/common/http/async_client_impl.cc [hunk 2]
```diff
@@ -87,6 +90,18 @@ void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
...
   ENVOY_LOG(trace, "async http request response data (length={} end_stream={})", data.length(),
             end_stream);
+  // It's possible we have been reset by the client but the router is still
+  // sending data on its way out. This happens for example when
+  // Utility::sendLocalReply() tells the gRPC async client that it has a non-200
+  // via encodeHeaders(), gRPC async client resets the HTTP async stream but on
+  // return to Utility::sendLocalReply() it unconditionally sends a body with
+  // encodeData().
+  // TODO(htuch): This is the only case we catch today, but conceivably we could
+  // see encodeTrailers() as well (or encodeHeaders()?) from router on its way
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: Per above I think we need to guard encodeTrailers() if we do the fix like this. I don't understand the comment about encodeHeaders() ? You guard that above, but I don't think that can happen?

In general, I agree this is sketchy. Arguably, this is a bug in Utility::secondLocalReply(). Basically, even local replies should adhere to a reset and stop sending data. Fixing that is probably out of scope for this change, but I might put in a TODO/NOTE that we should look at this more.

re: coverage you should be able to hit this via unit tests I think?

**Comment by msample** [comment]
data point: new async client unary send works great for my filter so long as gRPC injection service can be connected to.  As soon that connection cannot be established (was TLS issue), it bails at the assertion (even with the 3 async client changes in this PR)

**Comment by mattklein123** [comment]
Yup @htuch is fixing. If you give us a few more days all of this code will be running in prod at Lyft and we can say with reasonable assurance that it works.

**Comment by mattklein123** [comment]
Hmm sorry read that too quickly @msample. If it still fails with this PR there may be some other issue still.

File: source/common/http/async_client_impl.cc (hunk 1 above)
Comment (htuch) [inline source/common/http/async_client_impl.cc]: Sorry, this PR is stale, I think I had some dirty client state locally that had remove from `encode{Headers,Trailers}` the guard.

File: source/common/http/async_client_impl.cc (hunk 2 above)
Comment (htuch) [inline source/common/http/async_client_impl.cc]: Yeah, I think we should ship the fix and test, will add a `TODO` and look into fixing the broader problem today, it seems dodgy to leave it for much longer. 

For coverage, to do unit tests I need to make the stream decoder callbacks on `Http::AsycnStreamImpl` public to poke at via the unit test. I notice we do this in various places, seems not so great though to have to change visibility of entire interfaces just for unit tests.

**Comment by htuch** [comment]
@msample I'll add an integration test for connection failure today as well, hopefully we can trigger this behavior as well in test.

File: source/common/http/async_client_impl.cc (hunk 2 above)
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: re: testing, I don't follow. In the tests you control the async client that gRPC async client uses. You can mock this and do whatever you want? (Fire callbacks on the decoder, etc.). I think you can do whatever you want without any other changes? (We have lots of tests like this).

File: source/common/http/async_client_impl.cc (hunk 1 above)
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: IMO we want a guard in data/trailers, but not headers. I don't think there is any reasonable way headers can happen after reset with a substantial bug beyond what we have today.

**Comment by htuch** [comment]
It's the HTTP client the needs the unit tests, I'll put out an update when
I get to the office to demonstrate (with callback visibility changes).

**Comment by mattklein123** [comment]
Ah I see. Ok. If it can't happen it can't happen. Local reply can only do
this with data so I would just fix that until we come up with a better
solution.

**Comment by htuch** [comment]
Actually, with suitable downcasting and upcasting, we can avoid visibility changes on the callbacks to do unit tests.

**Comment by htuch** [comment]
@msample I just added two integration tests for initial connection issues, I don't see the assertion triggered. Do you think there is some code path that isn't being exercised here?

**Comment by msample** [comment]
@htuch I will give it another try with the 3 async_client_impl.cc changes in this PR changes later today.  In my filter's case, if it can't talk to the gRPC service to get headers, it lets the original request continue upstream as-is. That may be a different flow than your test.

**Comment by mattklein123** [comment]
Per offline convo given simplicity I would be in favor of just fixing https://github.com/lyft/envoy/issues/1283 when fixing this. But I can go either way.

**Pull request merged**